odoo/tools/convert.py

```diff
@@ -736,32 +721,34 @@ def model_id_get(self, id_str, raise_if_not_found=True):
             id_str = '%s.%s' % (self.module, id_str)
         return self.env['ir.model.data'].xmlid_to_res_model_res_id(id_str, raise_if_not_found=raise_if_not_found)
 
-    def parse(self, de, mode=None):
-        roots = ['openerp','data','odoo']
-        if de.tag not in roots:
-            raise Exception("Root xml tag must be <openerp>, <odoo> or <data>.")
-        for rec in de:
-            if rec.tag in roots:
-                self.parse(rec, mode)
-            elif rec.tag in self._tags:
-                try:
-                    self._tags[rec.tag](rec, de, mode=mode)
-                except Exception as e:
-                    exc_info = sys.exc_info()
-                    pycompat.reraise(
-                        ParseError,
-                        ParseError(ustr(e), etree.tostring(rec, encoding='unicode').rstrip(), rec.getroottree().docinfo.URL, rec.sourceline),
-                        exc_info[2]
-                    )
-        return True
+    def _tag_root(self, el, _data_node):
+        for rec in el:
+            f = self._tags.get(rec.tag)
+            if f is None:
+                continue
+
+            self.envs.append(self.get_env(el))
+            try:
+                f(rec, el)
+            except Exception as e:
+                exc_info = sys.exc_info()
+                pycompat.reraise(
+                    ParseError,
+                    ParseError(ustr(e), etree.tostring(rec, encoding='unicode').rstrip(), rec.getroottree().docinfo.URL, rec.sourceline),
+                    exc_info[2]
+                )
```

Comment:
This may lead to recursive wrapping with ParseError().
The exception handling should be at the toplevel instead (in `parse`).

Reply:
Good point, should be fixed.